### PR TITLE
#5 - Handling the expiration refresh

### PR DESCRIPTION
**Pull request by lcobucci**

### CODE REVIEW

**Comment by Ocramius** [comment]
@lcobucci I applied fixes to make this patch green again

**Comment by Ocramius** [comment]
Linking #5

**Comment by Ocramius** [comment]
I pushed up branch [`ookie-refresh-with-lazy-sessions`](https://github.com/Ocramius/StorageLessSession/tree/cookie-refresh-with-lazy-sessions), which is rebased on the lazy sessions stuff (works)

File: src/StoragelessSession/Http/SessionMiddleware.php [hunk 1]
```diff
@@ -234,13 +237,26 @@ private function appendToken(SessionInterface $sessionContainer, Response $respo
...
+     * {@inheritDoc}
+     */
+    private function shouldTokenBeRefreshed(Token $token = null) : bool
+    {
+        if (null === $token) {
+            return false;
         }
 
-        return FigResponseCookies::set($response, $this->getTokenCookie($sessionContainer));
+        return time() >= ($token->getClaim('exp') - $this->refreshTime);
```
Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: As discussed with @malukenho, this expression is not fully true.

To make this more practical:

`<token-created>=========<token-refresh>=========================================<token-expires>`

The time between `token-created` and `token-expires` is `expirationTime`, but what do we name the time between `token-created` and `token-refresh`? 
Also, is this operation valid right now? Or are we just subtracting (by default) 60 seconds from the end?

Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Should probably be `return time() >= (($token->getClaim('exp') - $this->expirationTime) + $this->refreshTime);`, which in practical terms (with the defaults) means "60 seconds after the token was created"

Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Maybe we should really base ourselves on the "issuedAt" claim, so the naming doesn't need to change, and it's actually clearer ("we refresh the token every 60 seconds" - quite easy to understand).

```php
return time() >= ($token->getClaim('iat') + $this->refreshTime);
```

Comment (lcobucci) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Yeah, that's easier.

File: src/StoragelessSession/Http/SessionMiddleware.php [hunk 2]
```diff
@@ -234,13 +237,26 @@ private function appendToken(SessionInterface $sessionContainer, Response $respo
...
+     * {@inheritDoc}
+     */
+    private function shouldTokenBeRefreshed(Token $token = null) : bool
+    {
+        if (null === $token) {
+            return false;
         }
 
-        return FigResponseCookies::set($response, $this->getTokenCookie($sessionContainer));
+        return time() >= ($token->getClaim('iat') + $this->refreshTime);
```
Comment (lcobucci) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Now our problem lies on the ```testWillIgnoreUnSignedTokens``` case.

We can change the claim values but that's just a distraction.
What is really happening is that we're considering an unsigned token - or a token with invalid signature - a valid thing to be used, and with that we're appending the token cookie (and returning a new response instance).

A nasty thing that can be done is to use references to invalidate the token instance when the signature is invalid, but if the callback that creates the ```SessionInterface``` is not called we'll probably append the cookie.

Any thoughts?

Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: What are the possible attack vectors there?

File: src/StoragelessSession/Http/SessionMiddleware.php [hunk 3]
```diff
@@ -193,54 +203,61 @@ private function parseToken(Request $request)
...
      *
      * @return Response
      *
      * @throws \InvalidArgumentException
      */
-    private function appendToken(SessionInterface $sessionContainer, Response $response) : Response
+    private function appendToken(SessionInterface $sessionContainer, Response $response, Token $token = null) : Response
     {
         $sessionContainerChanged = $sessionContainer->hasChanged();
+        $sessionContainerIsEmpty = $sessionContainer->isEmpty();
```
Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Fetching this value eagerly is not possible - if you do that, then you are initializing the session container each time

File: src/StoragelessSession/Http/SessionMiddleware.php (hunk 2 above)
Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: As discussed, the only possible outcome of this is that the user will trigger crypto token verification at each page load. Unless the user is able to forge a pseudo-real signature, this process will bail out early and have no DDoS-alike effects anyway.

**Comment by Ocramius** [comment]
Apparently, PHP7 builds on travis are super-broken now.

File: src/StoragelessSession/Http/SessionMiddleware.php [hunk 4]
```diff
@@ -85,7 +94,8 @@ public function __construct(
         string $verificationKey,
         SetCookie $defaultCookie,
         Parser $tokenParser,
-        int $expirationTime
+        int $expirationTime,
+        int $refreshTime = self::DEFAULT_REFRESH_TIME
```
Comment (Ocramius) [inline src/StoragelessSession/Http/SessionMiddleware.php]: Not really happy with this one - maybe make it mandatory, then set the default only when using the "helper constructors"?

**Comment by Ocramius** [comment]
Still needs a test that checks the refresh interval precisely (and/or uses a custom refresh interval)

**Comment by Ocramius** [comment]
Hmm, no, already handled by `testShouldNotRegenerateTokenWhenRequestHasATokenThatIsFarFromExpiration`

**Pull request merged**